blog/content/posts/client_registration.md

```diff
@@ -0,0 +1,165 @@
+
+```mermaid
+sequenceDiagram
+   participant Client
+   participant User
+   participant AuthServer
+   participant MetadataURL
+
+   Client->>User: Start OAuth flow
+   User->>AuthServer: Authorization request<br/>(client_id=https://app.com/oauth.json)
+   AuthServer->>MetadataURL: GET https://app.com/oauth.json
+   MetadataURL-->>AuthServer: {name: "App", redirect_uris: [...]}
+   AuthServer->>User: Show consent screen<br/>("App wants access...")
+   User->>AuthServer: Approve
+   AuthServer-->>Client: Authorization code
+   Client->>AuthServer: Exchange code for token<br/>(client_id=https://app.com/oauth.json)
+   AuthServer-->>Client: Access token
+```
+
+Checking against our goals:
+* The servers can trust the client metadata by trusting the domain the metadata is hosted on
+* Servers will have a single client ID per client (the metadata URL)
+* Servers can allow or deny clients based on their metadata URL or metadata domain
+* Servers don't need to handle any database by default, as they can fetch the metadata at authorization time.
+
+The costs for this implementation are:
+* Clients need to host a metadata document on an https URL
+* Clients must provide the URL as their client ID (if they determine an authorization server supports this)
+* Servers need to fetch the metadata, opening egress on their authorization request
+
+Client ID metadata can also be extended to require a JWKS and signed attestation at request time if desired.
+
+This is discussed more in this SEP: (TODO: link to SEP).
+
+
+## Comparison of Client Registration Approaches
+
+| Approach | Client Developer Effort | User Effort | Server Trust in Metadata | Single Client ID per App | Selective Allow/Deny | Database Management |
+|----------|------------------------|-------------|-------------------------|-------------------------|---------------------|-------------------|
+| **Pre-registration (by developer)** | ❌ Must coordinate with each server | ✅ None - just connects | ✅ Full trust via manual verification | ✅ Yes | ✅ Full control | ✅ Simple, bounded |
+| **Pre-registration (by user)** | ✅ None | ❌ Must register & manage credentials | ✅ User explicitly registers | ❌ Each user gets own ID | ✅ Full control | ⚠️ Grows with users |
+| **Basic DCR** | ✅ None | ✅ None | ❌ Must trust self-asserted data | ❌ New ID per user/session | ⚠️ Limited (blocklists) | ❌ Unbounded growth or complex expiry |
+| **DCR + Software Statements** | ⚠️ Must host JWKS & sign statements | ✅ None | ✅ Trust via domain/signature | ✅ Possible (via key identity) | ✅ Via trusted domains | ⚠️ Better but still needs storage |
+| **Client ID Metadata** | ⚠️ Must host metadata document | ✅ None | ✅ Trust via domain | ✅ Yes (URL is ID) | ✅ Via domains | ✅ No storage needed* |
+
+\* Servers may choose to cache for performance but aren't required to maintain a registration database
```

Comment:
While no registration database is _required_, trust conscious implementations will need metadata caching, fingerprinting for change detection, and policy enforcement. When metadata changes (client name, logo, redirect URIs, etc.), servers must decide whether existing user consents remain valid, since users made decisions based on different information.